scripts/get_maintainer.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2019, Linaro Limited
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+from pathlib import PurePath
+from urllib.request import urlopen
+
+import argparse
+import glob
+import os
+import re
+import tempfile
+
+
+DIFF_GIT_RE = re.compile(r'^diff --git a/(?P<path>.*) ')
+REVIEWED_RE = re.compile(r'^Reviewed-by: (?P<approver>.*>)')
+ACKED_RE = re.compile(r'^Acked-by: (?P<approver>.*>)')
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='Print the maintainers for '
+                                     'the given source files or directories; '
+                                     'or for the files modified by a patch or '
+                                     'a pull request. '
+                                     '(With -m) Check if a patch or pull '
+                                     'request is properly Acked/Reviewed for '
+                                     'merging.')
+    parser.add_argument('-m', '--merge-check', action='store_true',
+                        help='use Reviewed-by: and Acked-by: tags found in '
+                        'patches to prevent display of information for all '
+                        'the approved paths.')
+    parser.add_argument('-p', '--show-paths', action='store_true',
+                        help='show all paths that are not approved.')
+    parser.add_argument('-s', '--strict', action='store_true',
+                        help='stricter conditions for patch approval check: '
+                        'subsystem "THE REST" is ignored for paths that '
+                        'match some other subsystem.')
+    parser.add_argument('arg', nargs='*', help='file or patch')
+    parser.add_argument('-f', '--file', action='append',
+                        help='treat following argument as a file path, not '
+                        'a patch.')
+    parser.add_argument('-g', '--github-pr', action='append', type=int,
+                        help='Github pull request ID. The script will '
+                        'download the patchset from Github to a temporary '
+                        'file and process it.')
+    return parser.parse_args()
+
+
+# Parse MAINTAINERS and return a dictionary of subsystems such as:
+# {'Subsystem name': {'R': ['foo', 'bar'], 'S': ['Maintained'],
+#                     'F': [ 'path1', 'path2' ]}, ...}
+def parse_maintainers():
+    subsystems = {}
+    with open("MAINTAINERS", "r") as f:
```

Comment:
This implies that you need to run the script from the root of the optee_os folder? Could we check both from the root and from the script folder itself? I.e., let the script figure it out by itself by testing both? I did something like that [here](https://github.com/Linaro/jipdate/blob/master/cfg.py#L73-L87) for Jipdate when looking for config files.

Reply:
Yes, good idea. Perhaps the most universal way would be to get the script path with `os.path.realpath(__file__)` and go up one level from there. I'll update.

Reply:
Yeah, that sounds like a good solution.

Reply:
Second thought: if we allow running `get_maintainer.py` from anywhere, then it could be error prone because the paths expected on the command line are relative to the project root.

For istance, `./scripts/get_maintainer.py README.md` designates the `README.md` file in the top directory. If we run the script from withing `scripts/`, then how should the file be specified? First option is `../README.md`, but doing so means doing some fragile path translation work. Second option is `README.md`, which could be misleading (what if `scripts/` contains a `README.md`?) and is inconvenient since you can't use shell completion to enter the file path.

I'd say we should just error out if MAINTAINERS is not in the current directory. The Linux `get_maintainer.pl` does [a similar thing](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/scripts/get_maintainer.pl?h=v5.3-rc4#n328) although I do quite get how one can use `--no-tree`.

Reply:
Another way would simply to check what path you were running the script from. If it's anything but `os.path.realpath(__file__)/..` then output an error message saying that it should be run from the top. I.e. disallow any paths that are not the top one (works as long as we don't want to use this script for the other optee_xyz gits).